### ISSUE TITLE
GPU and CPU from util 100% to util 0% after 5-6 epochs. 

### ISSUE THREAD

**Issue by Nicolaas94** [MIC-DKFZ/nnUNet#2041]
First of all thank you very much for this great platform with very nice documentation! 
Unfortunately, I am running in a idle problem after the first 5-6 epochs. 
Could you please help me out? I have listed as much as information possible below.

I am running a virtual GPU with Ubuntu;
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 18.04.5 LTS
Release:        18.04
Codename:       bionic
with Grid T4-8Q profile 8192MiB

I create a new virtual environment with anaconda on this virtual machine. The datasets are stored on this machine as well.  
{    "name": "Task",
    "description": "SegMed dataset converted to MSD format",
    "reference": "Empty",
    "license": "Empty",
    "release": "Empty",
    "labels": {
             "background": 0,
             "mask_s": [1, 2],
             "mask_t": 2
    },
    "regions_class_order": [1, 2],
	     "numTraining": 89,
    "channel_names": {
             "0": "US"
    },
    "file_ending": ".nii.gz"}

First I installed pytorch following their website: `conda install pytorch torchvision torchaudio pytorch-cuda=12.1 -c pytorch -c nvidia`.
Then I installed: `pip install nnunetv2`. 

lscpu tells me:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              12
On-line CPU(s) list: 0-11
Thread(s) per core:  1
Core(s) per socket:  1
Socket(s):           12
NUMA node(s):        2
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) Gold 6146 CPU @ 3.20GHz
Stepping:            0
CPU MHz:             3192.499
BogoMIPS:            6384.99
Hypervisor vendor:   VMware
Virtualization type: full
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            25344K
NUMA node0 CPU(s):   0-5
NUMA node1 CPU(s):   6-11

GPU during training shows:
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.104.05             Driver Version: 535.104.05   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  GRID T4-8Q                     On  | 00000000:02:00.0 Off |                  N/A |
| N/A   N/A    P0              N/A /  N/A |   7032MiB /  8192MiB |     97%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+

+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A      2919      C   ...miniconda3/envs/nnunetv2/bin/python     7032MiB |
**+---------------------------------------------------------------------------------------+w**

and the `top` command during training shows:
multipe processes of `nnUnNtv2_train` of which 1 is around 100% or more. 
and the other processes are around 50%. 
I/O bottleneck is excluded as wa= 0,0

But after 5 or 6 epochs. 
the cpu processes reduce to max 1%, and the GPU util max 1%. 
I am the only user of this virtual machine, without running other programs. 
The cpu does not show any other software using the its util for >1%. 

What happens during training that stops the whole process, while this is not a problem during the first epochs?

I have tried the following stuff based on your documentation and other issues but unfortunately did not solve the problem:
- Again, create a venv and install the pytorch first, followed by the `pip install nnunetv2`. 
- To tackle a CPU bottleneck I added, as advised in the benchmark document, the line: `export nnUNet_n_proc_DA=32` and `export OMP_NUM_THREADS=1`
- To tackle the I/O bottleneck, the `wa =0,0`, so it doesn't seem to be the problem. 
- in simpleitk_reader_writer.py I changed the float32 to float16 as was advised in issue: "nnUnetv2 training freezes after the first epoch #1919" and redo `nnUNetv2_plan_and_preprocess`

verbose shows me:
2024-03-26 09:03:27.745975: unpacking dataset...
2024-03-26 09:03:30.885111: unpacking done...
2024-03-26 09:03:30.895031: do_dummy_2d_data_aug: False
2024-03-26 09:03:30.896398: Using splits from existing split file: .../splits_final.json
2024-03-26 09:03:30.897185: The split file contains 5 splits.
2024-03-26 09:03:30.897259: Desired fold for training: 0
2024-03-26 09:03:30.897306: This split has 71 training and 18 validation cases.
2024-03-26 09:03:30.915425: Unable to plot network architecture:
2024-03-26 09:03:30.915524: No module named 'IPython'
2024-03-26 09:03:30.943827:
2024-03-26 09:03:30.943943: Epoch 0
2024-03-26 09:03:30.944135: Current learning rate: 0.01
using pin_memory on device 0
using pin_memory on device 0
2024-03-26 09:06:26.123433: train_loss -0.1781
2024-03-26 09:06:26.123756: val_loss -0.4424
2024-03-26 09:06:26.123846: Pseudo dice [0.888, 0.0]
2024-03-26 09:06:26.123935: Epoch time: 175.18 s
2024-03-26 09:06:26.124009: Yayy! New best EMA pseudo Dice: 0.444
2024-03-26 09:06:27.577732:
2024-03-26 09:06:27.578282: Epoch 1
2024-03-26 09:06:27.578514: Current learning rate: 0.00999
2024-03-26 09:09:13.552649: train_loss -0.5918
2024-03-26 09:09:13.553180: val_loss -0.7169
2024-03-26 09:09:13.553325: Pseudo dice [0.9318, 0.5979]
2024-03-26 09:09:13.553442: Epoch time: 165.98 s
2024-03-26 09:09:13.553530: Yayy! New best EMA pseudo Dice: 0.4761
2024-03-26 09:09:15.566862:
2024-03-26 09:09:15.567165: Epoch 2
2024-03-26 09:09:15.567363: Current learning rate: 0.00998
2024-03-26 09:11:56.184359: train_loss -0.794
2024-03-26 09:11:56.184972: val_loss -0.76
2024-03-26 09:11:56.185201: Pseudo dice [0.9423, 0.6545]
2024-03-26 09:11:56.185370: Epoch time: 160.62 s
2024-03-26 09:11:56.185514: Yayy! New best EMA pseudo Dice: 0.5083
2024-03-26 09:11:58.150763:
2024-03-26 09:11:58.150897: Epoch 3
2024-03-26 09:11:58.150990: Current learning rate: 0.00997
2024-03-26 09:14:42.860391: train_loss -0.8333
2024-03-26 09:14:42.860705: val_loss -0.7672
2024-03-26 09:14:42.860818: Pseudo dice [0.9437, 0.6588]
2024-03-26 09:14:42.860898: Epoch time: 164.71 s
2024-03-26 09:14:42.860955: Yayy! New best EMA pseudo Dice: 0.5376
2024-03-26 09:14:44.967245:
2024-03-26 09:14:44.967527: Epoch 4
2024-03-26 09:14:44.967663: Current learning rate: 0.00996
2024-03-26 09:17:27.299671: train_loss -0.8517
2024-03-26 09:17:27.299984: val_loss -0.7676
2024-03-26 09:17:27.300065: Pseudo dice [0.9518, 0.6446]
2024-03-26 09:17:27.300146: Epoch time: 162.33 s
2024-03-26 09:17:27.300202: Yayy! New best EMA pseudo Dice: 0.5637
2024-03-26 09:17:29.395443:
2024-03-26 09:17:29.395597: Epoch 5
2024-03-26 09:17:29.395701: Current learning rate: 0.00995
And then this epoch will take around 25000 seconds. 

Benchmark shows: 
cudnn_version: 8902
GPU_name: GRID T4-8Q
torch_version: 2.2.1

Dataset | Trainer | Config| Fastest_epoch
--          | -- | -- | --
2_heart | nnUNetTrainerBenchmark_5epochs                           | 2d | 158,3
2           | nnUNetTrainerBenchmark_5epochs                           | 3d_fullres | 220,883
2           | nnUNetTrainerBenchmark_5epochs_noDataLoading | 2d | 157,78
2           | nnUNetTrainerBenchmark_5epochs_noDataLoading | 3d_fullres | 220,805
  |   |   |  
4_hippocampus  | nnUNetTrainerBenchmark_5epochs                           | 2d | 44,91
4                         | nnUNetTrainerBenchmark_5epochs                           | 3d_fullres | 32,55
4                         | nnUNetTrainerBenchmark_5epochs_noDataLoading | 2d | 39,39
4                         | nnUNetTrainerBenchmark_5epochs_noDataLoading | 3d_fullres | 31,65

I would really appreciate your help or any advice. 

Kind regards, 
Nicolaas

**Comment by CodeHarcourt** [comment]
Are your train_loss and val_loss Synchronous ascent? You just train 5 epochs. Can you show more epoch results? When I train 50 epochs I find that the train_loss and val_loss Synchronous ascent? Aren't you?

**Comment by Nicolaas94** [comment]
> Are your train_loss and val_loss Synchronous ascent? You just train 5 epochs. Can you show more epoch results? When I train 50 epochs I find that the train_loss and val_loss Synchronous ascent? Aren't you?

Well, after 5 epochs, the system stops, and training takes too long. So then I quit training.

**Comment by Nicolaas94** [comment]
> 1. RAM: How much RAM do you have? Did you check your RAM usage? Did you somehow started to use the swap memory?
> 2. Shared memory: What is the shared memory size of your virtual machine? Pytorch needs around 32 GB of memory.

1. the `free -m` command shows me this while training is still 'running'              
            |         total     |   used     |     free      |    shared    |  buff/cache  | available |
Mem:   |       48292    |    12748  |     32234   |   1200       |     3309         |   33885   |
Swap:   |        979       |      0        |      979     |

and `ipcs` results in an empty table: 
------ Message Queues --------
key        msqid      owner      perms      used-bytes   messages

------ Shared Memory Segments --------
key        shmid      owner      perms      bytes      nattch     status

------ Semaphore Arrays --------
key        semid      owner      perms      nsems


Do you thing this is still a RAM problem?

**Comment by ancestor-mithril** [comment]
I don't have any idea. I would recommend trying with `nnUNet_n_proc_DA=6` and monitoring both RAM and CPU.

**Comment by Nicolaas94** [comment]
I change it to `nnUNet_n_proc_DA=6`.
Then started training with the benchmark `Dataset004_Hippocampus`. 
Now GPU memory-Usage is 1570MiB /  8192MiB with 80% Util. 
But after 20 epochs the same problem occurs and both CPU and GPU Util drop to <1%. I still had lost of free memory left.

**Comment by ancestor-mithril** [comment]
If you reduce `nnUNet_n_proc_DA=6` even more, to `nnUNet_n_proc_DA=2`, does your training run longer? If so, this might really be a RAM issue. 
One thing you could try is to use the latest nnUNet from the `master` branch and check if the behavior persists.

**Comment by Nicolaas94** [comment]
> If you reduce `nnUNet_n_proc_DA=6` even more, to `nnUNet_n_proc_DA=2`, does your training run longer? If so, this might really be a RAM issue. One thing you could try is to use the latest nnUNet from the `master` branch and check if the behavior persists.

I reduced `nnUNet_n_proc_DA=2` but now only 13epochs were trained until CPU and GPU are idle. So that would mean it is not a RAM issue. 

Could you explain to me how to install the latest nnUNet from the master branch? is that the second option in the installation_instruction?
2.ii. : 
```
git clone https://github.com/MIC-DKFZ/nnUNet.git
cd nnUNet
pip install -e .
```

**Comment by ancestor-mithril** [comment]
Yes.

**Comment by CodeHarcourt** [comment]
Have you solved this problem? Can you show me your train_loss and val_loss?

**Comment by Nicolaas94** [comment]
I did `pip uninstall nnunetv2` and install it from the `master` branch 2.ii. reboot the system just to be sure. 
Unfortunately, the system still breaks down after 11 epochs (17 minutes) trying to train the benchmark Dataset004_Hippocampus.

**Comment by Nicolaas94** [comment]
I repeated the installation on a physical workstation, and now everything works fine. 
This workstation has more GPU and CPU power, which could be the problem for the virtual GPU. Or the virtual GPU with the GRID driver is incompatible with the nnUNetv2. For now, I can continue my research, but would be interesting to know for future research.

**Comment by FabianIsensee** [comment]
Hey, sorry for the late reply. Issues like these are super difficult for me to debug because it only appears on some systems. If there was a code snippet that could reproduce this then I could really dive deep into this and fix it.
Python multiprocessing is unfortunately always a big issue, but the most recent master should actually have this under control. We constantly monitor the background processes and notice when some are missing. This should alleviate most of the earlier problems that people had with insufficient RAM. Speaking of, this is clearly not the problem here.
I think some combination of the virtualization and multiprocessing could be at fault. In order to test this, use `nnUNet_n_proc_DA=0`. This will disable multiprocessing and run everything in the main process. Training will be super slow, so this isn't a viable solution but if this runs without interruptions we know that multiprocessing is to blame.
Another thing you could try is to set the multiprocessing start method to spawn (default in Ubuntu is fork and that is what the data augmentation uses).

Like I said, if you could manage to make this problem reproducible I can take a closer at it. The way it currently is I can unfortunately only guess :sweat_smile: 

Just FYI for regular training nnUNet_n_proc_DA should be 11 on your machine because you have 12 CPUs (11 for DA, one for the main process)

**Comment by FabianIsensee** [comment]
Hey, what's the status here?
